mutex/src/workflows.ts

```diff
@@ -0,0 +1,66 @@
+import * as workflow from '@temporalio/workflow';
+
+interface LockRequest {
+  initiatorId: string;
+  timeoutMs: number;
+}
+
+interface LockResponse {
+  releaseSignalName: string;
+}
+
+export const currentWorkflowIdQuery = workflow.defineQuery<string | null>('current-workflow-id');
+export const hasLockQuery = workflow.defineQuery<boolean>('hasLock');
+export const lockRequestSignal = workflow.defineSignal<[LockRequest]>('lock-requested');
+export const lockAcquiredSignal = workflow.defineSignal<[LockResponse]>('lock-acquired');
+
+export async function lockWorkflow(requests = Array<LockRequest>()) {
+  let currentWorkflowId: string | null = null;
+  workflow.setHandler(lockRequestSignal, (req: LockRequest) => {
+    requests.push(req);
+  });
+  workflow.setHandler(currentWorkflowIdQuery, () => currentWorkflowId);
+  while (workflow.workflowInfo().historyLength < 2000) {
+    await workflow.condition(() => requests.length > 0);
+    const req = requests.shift();
+    if (req === undefined) {
+      continue;
+    }
```

Comment:
when would this be true?

Reply:
Never. This was just to prevent TS from complaining that `req` is potentially undefined on subsequent lines.

Reply:
Yeah, TS can't know that the array is empty here.
You might want to add a comment explaining why this is here.

Reply:
is casting better?

```
const req = requests.shift() as LockRequest; // this will never be undefined
```